Sparrow.ps1

```diff
@@ -184,12 +187,20 @@ Function Get-UALData {
     #By default, it will show up as Consent_Operations_Export.csv       
     Export-UALData -ExportDir $ExportDir -UALInput $ConsentData -CsvName "Consent_Operations_Export" -WorkloadType "AAD"
 
-    #Searches for SAML token usage anomaly (UserAuthenticationValue of 16457) in the Unified Audit Logs
-    Write-Verbose "Searching for 16457 in UserLoggedIn and UserLoginFailed operations in the UAL."
-    $SAMLData = Search-UnifiedAuditLog -StartDate $StartDate -EndDate $EndDate -Operations "UserLoggedIn","UserLoginFailed" -ResultSize 5000 -FreeText "16457" | Select-Object -ExpandProperty AuditData | Convertfrom-Json
-    #You can modify the resultant CSV output by changing the -CsvName parameter
-    #By default, it will show up as SAMLToken_Operations_Export.csv      
-    Export-UALData -ExportDir $ExportDir -UALInput $SAMLData -CsvName "SAMLToken_Operations_Export" -WorkloadType "AAD"
+    #Searches for SAML token usage anomaly (UserAuthenticationValue of 16457) in the Unified Audit 
```

Comment:
Get-MsolDomainFederationSettings will fail for child federated domains. Here is a logic to check only root domains and add all child domains that chain up to a root that doesn't support federated MFA.


```suggestion
  #Searches for SAML token usage anomaly (UserAuthenticationValue of 16457) in the Unified Audit
  $federatedDomains = Get-MsolDomain | Where-Object {$_.Authentication -eq "Federated"}
  # Get only root domains so we can get SupportMFA status
  $rootDomains = $federatedDomains | Where-Object {$_.RootDomain -eq $null}
  # Get root domains that don't support MFA, hence Federated MFA is not expected. Note: federated MFA is still possible when SupportsMFA is false, however less likely. Check your STS configuration.
  $rootDomainsSupportMFAFalse = @()
  
  foreach ($rootDomain in $rootDomains)
  {
    $fedProps = Get-MsolDomainFederationSettings -DomainName $rootDomain.Name 
    If ($fedProps.SupportsMfa -eq $False) {
        $rootDomainsSupportMFAFalse += $rootDomain.Name
    }
  }
# Add all child domains where its root is on the list
$childDomainsSupportMFAFalse = @()
$childDomains = $federatedDomains | Where-Object {$_.RootDomain -ne $null}

foreach ($childDomain in $childDomains)
{
    if ($childDomain.RootDomain -in $rootDomainsSupportMFAFalse){
        $childDomainsSupportMFAFalse += $childDomain.name
    }
}

$domainsToFlag = $rootDomainsSupportMFAFalse + $childDomainsSupportMFAFalse
```